akka-http-core/src/main/scala/akka/http/scaladsl/model/ServerSentEvent.scala

```diff
@@ -0,0 +1,143 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package akka.http
+package scaladsl
+package model
+
+import akka.util.ByteString
+import java.nio.charset.StandardCharsets.UTF_8
+import scala.annotation.tailrec
+import scala.compat.java8.OptionConverters.RichOptionForJava8
+
+object ServerSentEvent {
+
+  /**
+   * A [[ServerSentEvent]] with empty data which can be used as a heartbeat.
+   */
+  val heartbeat: ServerSentEvent =
+    ServerSentEvent("")
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param type type, must not contain \n or \r
+   */
+  def apply(data: String, `type`: String): ServerSentEvent =
+    new ServerSentEvent(data, Some(`type`))
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param type type, must not contain \n or \r
+   * @param id id, must not contain \n or \r
+   */
+  def apply(data: String, `type`: String, id: String): ServerSentEvent =
+    new ServerSentEvent(data, Some(`type`), Some(id))
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param retry reconnection delay in milliseconds
+   */
+  def apply(data: String, retry: Int): ServerSentEvent =
+    new ServerSentEvent(data, retry = Some(retry))
+
+  private def noNewLine(s: String) = s.forall(c ⇒ c != '\n' && c != '\r')
+
+  // Public domain algorithm: http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2.
+  // We want powers of two both because they typically work better with the allocator, and because we want to minimize
+  // reallocations/buffer growth.
+  private def nextPowerOfTwoBiggerThan(n: Int) = {
+    var m = n - 1
+    m |= m >> 1
+    m |= m >> 2
+    m |= m >> 4
+    m |= m >> 8
+    m |= m >> 16
+    m + 1
+  }
+}
+
+/**
+ * Representation of a server-sent event. According to the specification, an empty data field designates an event
+ * which is to be ignored which is useful for heartbeats.
+ *
+ * @param data data, may span multiple lines
+ * @param eventType optional type, must not contain \n or \r
+ * @param id optional id, must not contain \n or \r
+ * @param retry optional reconnection delay in milliseconds
+ */
+final case class ServerSentEvent(
+  data:      String,
+  eventType: Option[String] = None,
+  id:        Option[String] = None,
+  retry:     Option[Int]    = None)
+  extends javadsl.model.ServerSentEvent {
+  import ServerSentEvent._
+
+  require(eventType.forall(noNewLine), "eventType must not contain \\n or \\r!")
```

Comment:
Nope. An empty event type is valid per specification, but in our nice typed Scala world semantically the same like `None`.